### PR TITLE
Add Vue Language Support

### PR DESCRIPTION
**Pull request by tylersatre**
Adds full Vue.js language support to Serena using the official `@vue/language-server`.

## Key Changes

**New Vue Language Server** (`src/solidlsp/language_servers/vue_language_server.py`)
- Integrates `@vue/language-server` for Vue SFC (`.vue`) file support
- Manages a companion TypeScript server for cross-file references between Vue components and TypeScript modules
- Handles proper language ID detection (vue vs typescript) for files within Vue projects
- Supports all standard LSP operations: symbol retrieval, go-to-definition, find references, and rename

**Symbol Resolution Fix** (`src/serena/code_editor.py`)
- Improved symbol disambiguation logic to handle Vue patterns like `defineExpose({ pressCount })` where the same symbol appears as both definition and shorthand property reference
- Selects the definition (larger range) when multiple candidates exist

**Configuration & Integration**
- Added `VUE` language enum and file matchers for `.vue` and TypeScript/JavaScript variants
- Added `_get_language_id_for_file()` hook in `SolidLanguageServer` for language-specific file type handling

## Tests

Comprehensive test suite covering:
- Basic LSP operations (symbols, definitions, references)
- Error handling and edge cases
- Cross-file rename operations
- Symbol retrieval in Vue SFCs and TypeScript files

Test repository includes Vue components, composables, Pinia stores, and TypeScript type definitions to validate real-world Vue project patterns.

### CODE REVIEW

File: test/solidlsp/vue/test_vue_rename.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        workspace_edit = language_server.request_rename_symbol_edit(file_path, sel_start["line"], sel_start["character"], "processDigit")
+
+        assert workspace_edit is not None, "Should return WorkspaceEdit for rename operation"
+
+        has_changes = "changes" in workspace_edit and workspace_edit["changes"]
+        has_document_changes = "documentChanges" in workspace_edit and workspace_edit["documentChanges"]
+
+        assert has_changes or has_document_changes, "WorkspaceEdit should contain either 'changes' or 'documentChanges'"
+
+        if has_changes:
```
Comment (MischaPanch) [inline test/solidlsp/vue/test_vue_rename.py:33]: should also handle the case of has_document_changes or remove the if - this whole block might never be called

File: test/solidlsp/vue/conftest.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import os
+import subprocess
+import time
+from pathlib import Path
+
+import pytest
+
+
+def ensure_vue_test_repo_dependencies(repo_path: str) -> None:
```
Comment (MischaPanch) [inline test/solidlsp/vue/conftest.py]: this should not be necessary, the LS should download the dependencies on the fly when it starts the first time. Most LS implementations in Serena do that. Also, pls make sure to never use emojis or prints (use log.info instead)

File: src/solidlsp/language_servers/typescript_language_server.py [hunk 3]
```diff
@@ -39,11 +40,34 @@ def getpwuid(uid: Any) -> Any:
...
+            return d
+    return definitions[0]
+
+
 class TypeScriptLanguageServer(SolidLanguageServer):
     """
     Provides TypeScript specific instantiation of the LanguageServer class. Contains various configurations and settings specific to TypeScript.
     """
 
+    # Version constants - subclasses can override these if needed
```
Comment (MischaPanch) [inline src/solidlsp/language_servers/typescript_language_server.py]: this should be set through `ls_specific_settings`

File: src/solidlsp/language_servers/vue_language_server.py [hunk 4]
```diff
@@ -0,0 +1,729 @@
...
+                new_item: dict = {}
+                new_item.update(item)  # type: ignore[arg-type]
+                new_item["absolutePath"] = str(abs_path)
+                new_item["relativePath"] = str(rel_path)
+                result.append(ls_types.Location(**new_item))  # type: ignore
+
+        return result
+
+    def request_file_references(self, relative_file_path: str) -> list:
+        from pathlib import Path
```
Comment (MischaPanch) [inline src/solidlsp/language_servers/vue_language_server.py]: pls prefer module-level imports

File: src/solidlsp/language_servers/vue_language_server.py [hunk 5]
```diff
@@ -0,0 +1,729 @@
...
+    def request_references(self, relative_file_path: str, line: int, column: int) -> list[ls_types.Location]:
+        if not self.server_started:
+            log.error("request_references called before Language Server started")
+            raise SolidLSPException("Language Server not started")
+
+        if not self._has_waited_for_cross_file_references:
+            sleep(self._get_wait_time_for_cross_file_referencing())
+            self._has_waited_for_cross_file_references = True
+
+        if self._ts_server is not None:
```
Comment (MischaPanch) [inline src/solidlsp/language_servers/vue_language_server.py]: shouldn't self._ts_server always be available?

**Review by MischaPanch — CHANGES_REQUESTED** [review]
Very nice PR, solid contribution, thank you!

I did a first shallow review, pls make sure tests pass in windows and dependencies are downloaded on the fly. When that's done I will review one more time in more detail.

One thing that would be very useful if you have the capacity to do that is to extract a more general pattern - a language server that uses another language server for cross-reference finding. This would also make it possible to implement svelte support without much hassle.

As it stands, we would get Vue support, but for svelte support the generalization would need to happen to keep the code clean, which would be a somewhat large refactoring. If you do have time, it would be great to address this already (you don't have to add the Svelte support yourself now, of course)

File: test/solidlsp/vue/conftest.py (hunk 2 above)
Comment (tylersatre) [reply test/solidlsp/vue/conftest.py]: Good call, I ended up removing conftest.py since it's not needed

File: src/solidlsp/language_servers/typescript_language_server.py (hunk 3 above)
Comment (tylersatre) [reply src/solidlsp/language_servers/typescript_language_server.py]: I switched this up and also updated the versions for the Vue language server in the same fashion

File: src/solidlsp/language_servers/vue_language_server.py (hunk 5 above)
Comment (tylersatre) [reply src/solidlsp/language_servers/vue_language_server.py]: You're right

File: README.md [hunk 6]
```diff
@@ -70,8 +70,35 @@ Language servers provide support for a wide range of programming languages.
 With Serena's LSP library, we provide **support for over 30 programming languages**, including
 AL, Bash, C#, C/C++, Clojure, Dart, Elixir, Elm, Erlang, Fortran, Go, Haskell, Java, Javascript, Julia, Kotlin, Lua, Markdown, Nix, Perl, PHP, Python, R, Ruby, Rust, Scala, Swift, TypeScript, YAML and Zig.
 
-> [!IMPORTANT]
-> Some languages require additional dependencies to be installed; see the [Language Support](https://oraios.github.io/serena/01-about/020_programming-languages.html) page for details.
```
Comment (opcode81) [inline README.md:74]: The conflict here was incorrectly resolved. This must remain; concrete language-related information was moved to docs/

File: src/serena/code_editor.py [hunk 7]
```diff
@@ -270,12 +270,41 @@ def _find_unique_symbol(self, name_path: str, relative_file_path: str) -> Langua
...
+            """
+            range_data = symbol.symbol_root.get("range", {})
+            start_line = range_data.get("start", {}).get("line", 0)
+            end_line = range_data.get("end", {}).get("line", 0)
+            start_char = range_data.get("start", {}).get("character", 0)
+            end_char = range_data.get("end", {}).get("character", 0)
+            return (end_line - start_line, end_char - start_char)
+
+        # Sort by range size descending - larger ranges are typically definitions
+        sorted_candidates = sorted(symbol_candidates, key=get_range_size, reverse=True)
```
Comment (opcode81) [inline src/serena/code_editor.py]: While this may work for Vue.js, it is not an acceptable way of making a choice when the name path is non-unique in general.

For example, in Java, if there is an overloaded method `doSomething` but the overload is not uniquely specified in the name path, we cannot just use the overload that has the largest definition. We need to raise an error, forcing the caller to disambiguate.

Is there any basis to decide whether the logic you added should apply?

Comment (tylersatre) [reply src/serena/code_editor.py]: This comes down to me not checking AI work close enough when I was working through this.  You're completely correct that it's not something that should be handled in a language agnostic file like this, so I've moved the logic for this - https://github.com/tylersatre/serena/commit/0de00210031efe030f6f401eec4364de55ab2963

File: README.md (hunk 6 above)
Comment (tylersatre) [reply README.md:74]: I've got this fixed up, I think it changed from the time I started this branch till now and I hadn't gone back through.

**Comment by tylersatre** [comment]
> One thing that would be very useful if you have the capacity to do that is to extract a more general pattern - a language server that uses another language server for cross-reference finding. This would also make it possible to implement svelte support without much hassle.
> 
> As it stands, we would get Vue support, but for svelte support the generalization would need to happen to keep the code clean, which would be a somewhat large refactoring. If you do have time, it would be great to address this already (you don't have to add the Svelte support yourself now, of course)

@MischaPanch I did go ahead and create another branch that extracts this to a more general pattern.  If you check out https://github.com/tylersatre/serena/pull/2 you'll need the changes compared to this branch, to make it easier to look at just the companion server changes instead of adding Vue + companion server all together.

Let me know what you think and if you like it I can open the PR to merge it back to here instead of just showing the changes against my branch.

**Comment by opcode81** [comment]
@tylersatre would you say we are ready for the merge now?

Once thing I thought was missing was an explicit tie-breaking strategy (which I added with the most recent commit). Previously the preference for TypeScript/JavaScript over Vue when a project contains no .vue files was basically coincidental.

**Comment by tylersatre** [comment]
> @tylersatre would you say we are ready for the merge now?

I think so!  Good catch and fix on the tie-breaking strategy.

**Comment by opcode81** [comment]
Great! The macOS failure is an unrelated random failure.
Merging ...

Thanks a lot, @tylersatre.

**Comment by opcode81** [comment]
@tylersatre I had to squash the changes for the merge.

I have recreated your `companion_server` branch based on the squased commit here:
https://github.com/oraios/serena/tree/companion_server
You can build on this branch for a future PR. I think it'll be worthwhile.

In the future, if you want us to merge your PRs without squashing, keep the history clean by using amend and interactive rebase, such that every commit represents a well-defined, meaningful addition.
